examples/pingpong/enclave/src/main.rs

```diff
@@ -104,13 +108,26 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     Ok(())
 }
 
-fn gas_fn(response: &EnclaveResponse) -> DefaultTxConfig {
-    if matches!(response, EnclaveResponse::Ping(_)) {
-        DefaultTxConfig {
+struct GasSimulator;
+
+#[async_trait::async_trait]
+impl GasProvider<EnclaveResponse, DefaultChainClient> for GasSimulator {
+    async fn gas_for_tx(
+        &self,
+        tx: &EnclaveResponse,
+        chain_client: &DefaultChainClient,
+        contract: &AccountId,
+    ) -> Result<DefaultTxConfig, anyhow::Error> {
+        let default_config = DefaultTxConfig {
             gas: 2000000,
             amount: "11000untrn".to_string(),
-        }
-    } else {
-        unreachable!()
+        };
+        let gas_info = chain_client
+            .simulate_tx(contract, tx.as_slice().iter(), default_config)
+            .await?;
+        Ok(DefaultTxConfig {
+            gas: gas_info.gas_used,
```

Comment:
You need to do a bit more here to be safe that the Tx has enough gas:
Gas is broken in cosmos :D
```suggestion
            gas: gas_info.gas_used * 1.3, // 1.3 is for 30% gas adjustment, overshooting is fine; undershooting isn't
            amount: "{gas_info.gas_used * 1.3 * BASE_GAS_PRICE}untrn".to_string()  // pseudocode; function is needed
```

Current base gas:
```shell
curl -X 'GET' \
  'https://neutron-api.polkachu.com/feemarket/v1/gas_price/untrn' \
  -H 'accept: application/json'

{
  "price": {
    "denom": "untrn",
    "amount": "0.005300000000000000"
  }
}
```

If you offer too litle "gas" the tx can fail out of gas.
If you offer to little "amount" the tx can fail because the amount of "gas" does not correspond to the amount of tokens being spend - you can always overshoot.


At the moment we are always paying 0.011 untrn per tx - that might be fine for now.

Do we need to implement this handler on `cycles-protocol`?

Reply:
I implemented the adjustment part in the cycles-protocol PR here -> https://github.com/informalsystems/cycles-protocol/blob/39a975ba17a9ee0e8edfacc27064a3bdb833446a/crates/quartz/enclave/src/utils.rs#L44-L50

> If you offer to little "amount" the tx can fail because the amount of "gas" does not correspond to the amount of tokens being spend - you can always overshoot.

Interesting! Didn't think about the amount. I think it will be better to implement this on the `cycles-protocol` side. 

Is it okay to hardcode `BASE_GAS_PRICE` as `0.0053` for now?